server/src/main/scala/org/http4s/server/middleware/Throttle.scala

```diff
@@ -0,0 +1,121 @@
+package org.http4s.server.middleware
+
+import org.http4s.{Http, Response, Status}
+import cats.data.Kleisli
+import cats.effect.{Clock, Sync}
+import cats.effect.concurrent.Ref
+import scala.concurrent.duration.FiniteDuration
+import cats.implicits._
+import java.util.concurrent.TimeUnit.NANOSECONDS
+import scala.concurrent.duration._
+
+/**
+  * Transform a service to reject any calls the go over a given rate.
+  */
+object Throttle {
+
+  sealed abstract class TokenAvailability extends Product with Serializable
+  case object TokenAvailable extends TokenAvailability
+  final case class TokenUnavailable(retryAfter: Option[FiniteDuration]) extends TokenAvailability
+
+  /**
+    * A token bucket for use with the [[Throttle]] middleware.  Consumers can take tokens which will be refilled over time.
+    * Implementations are required to provide their own refill mechanism.
+    *
+    * Possible implementations include a remote TokenBucket service to coordinate between different application instances.
+    */
+  trait TokenBucket[F[_]] {
+    def takeToken: F[TokenAvailability]
+  }
+
+  object TokenBucket {
+
+    /**
+      * Creates an in-memory [[TokenBucket]].
+      *
+      * @param capacity the number of tokens the bucket can hold and starts with.
+      * @param refillEvery the frequency with which to add another token if there is capacity spare.
+      * @return A task to create the [[TokenBucket]].
+      */
+    def local[F[_]](capacity: Int, refillEvery: FiniteDuration)(
+        implicit F: Sync[F],
+        clock: Clock[F]): F[TokenBucket[F]] = {
+
+      def getTime = clock.monotonic(NANOSECONDS)
+      val bucket = getTime.flatMap(time => Ref[F].of((capacity.toDouble, time)))
```

Comment:
why is this a `Double` over a `Long`? You'll essentially be working with integral amounts most of the time.

Reply:
If we want to allow fractional tokens, which is the basis of the algorithm suggested by @rossabaker then I think we need a double? Unless I am missing something?